ViT UNETR SSL

valosekj opened the pull request:
Just a draft PR to allow commit suggestions

valosekj committed:
Add notebook debugging SSL pretraining transformations

valosekj committed:
Fix typo

valosekj committed:
Add script for Self-Supervised Pre-training using Vision Transformer (ViT)

valosekj committed:
Fix typo

valosekj committed:
Do sanity dimension check after transformations

valosekj committed:
Use `patch_size=(16, 16, 16)` and `batch_size = 2`

- `patch_size=(16, 16, 16)` - to fix tensor size mismatch
- `batch_size = 2` - to handle memory

valosekj committed:
rename 'crop_size' to 'spatial_size'

valosekj committed:
use 'contrast-agnostic' data augmentation for Training of the fine-tuned model

valosekj committed:
Add script to finetune the 3D Single-Class Spinal Cord Lesion Segmentation Model Using SSL Pre-trained Weights

valosekj committed:
update description

valosekj committed:
use 'logger.info' instead of 'print'

valosekj committed:
make '--data' arg notrequired

valosekj committed:
explicitly specify CUDA GPU number

valosekj committed:
change sliding_window_inference roi_size and batch_size

valosekj committed:
remove 'to_onehot_y=True' and 'softmax=True' because we work with single channel prediction

valosekj committed:
add notebook to debug transformations when using SC masks to crop patches

valosekj committed:
rerun the notebook to show different slices

valosekj committed:
Add 'keys' arg

valosekj committed:
Add 'keys' arg

valosekj committed:
Update 'RandCoarseDropoutd' params

valosekj committed:
Update comments

valosekj committed:
Fix imports

valosekj committed:
Change batch_size and NUM_WORKERS to 4

valosekj committed:
Make '--data' non required

valosekj committed:
change 'num_workers' to 0 to prevent 'RuntimeError: received 0 items of ancdata'

valosekj committed:
Update input arg description

valosekj committed:
Rerun the notebook

valosekj committed:
Print hyper-parameters into the log file

valosekj committed:
Add '--cuda' input arg

valosekj committed:
track and save epoch time

valosekj committed:
Plot and save input and output validation images to see how the model is learning

valosekj committed:
Add comment

valosekj committed:
Do not plot 'outputs_v2' as it is a hidden representation

valosekj committed:
Include the epoch number as master title

valosekj committed:
Add 'torch.multiprocessing.set_sharing_strategy('file_system')' to solve number of workers issue

Fixes: https://github.com/ivadomed/model-seg-dcm/issues/8

valosekj committed:
Create validation_figures directory if it does not exist

valosekj committed:
Use 3 leading zeros for the epoch number in the figures fname

valosekj committed:
Link issue

valosekj committed:
Add note for 'RandCropByPosNegLabeld'

It seems that the transform randomly selects a foreground point from image, then use it as center crop. This means that it can find the closest voxel that is just outside the SC and use it as the center (hence it includes the SC).
Source: https://github.com/Project-MONAI/MONAI/issues/452#issuecomment-636065502

valosekj committed:
Add 'number_of_holes' arg to specify the number of holes to be used for the RandCoarseDropoutd and RandCoarseShuffled transforms

valosekj committed:
typo

valosekj committed:
batch_size = 8

valosekj committed:
NUM_WORKERS = batch_size

valosekj committed:
number_of_holes=5

number of holes to be used for the RandCoarseDropoutd and RandCoarseShuffled transforms

naga-karthik commented on vit_unetr_ssl/transforms.py:
```diff
@@ -0,0 +1,224 @@
+            OneOf(
+                transforms=[
+                    # Randomly drop regions of the image
+                    # 'dropout_holes=True': the dropped regions will be set to zero, introducing regions of no
+                    # information within the image.
```
so I think this definition is incorrect -- if `dropout_holes=True` then the `holes` will be dropped out and the regions in the image with dropped holes will be filled with `fill_value` (i.e. they are not always set to zero, depends on `fill_value`)

valosekj replied:
Good catch! You are right!

From [RandCoarseDropoutd](https://docs.monai.io/en/latest/transforms.html#randcoarsedropoutd) documentation:

> dropout_holes – if True, dropout the regions of holes and fill value, if False, keep the holes and dropout the outside and fill value. default to True.

Since, I do not specify `fill_value`, `None` is used, meaning:

> if None, will compute the min and max value of input image then randomly select value to fill, default to None.

I further debugged and plotted the transforms. It confirms your findings! 

<img width="626" alt="image" src="https://github.com/ivadomed/model-seg-dcm/assets/39456460/20346e03-e046-4bee-8588-a7389117735b">

naga-karthik replied:
So regarding this -- I think it's best to set `fill_value=0`. Reason -- when the dropped out holes are randomly filled with values, then this makes it difficult for the model to learn (i.e. it is trying to learn that "there is a patch of random values, I have to figure out how to understand what's behind them"). 
INSTEAD, if we set `fill_value=0` then we're making the task simpler -- the models sees only black regions and it knows that these patches are uniform (i.e. filled only with zeros) instead of seeing that there are `n` holes and each hole has different values of pixels)

Bottomline -- set `fill_value=0` to make the training objective easier

valosekj replied:
Thanks for this suggestion! Makes sense! Trying `fill_value=0` now (https://github.com/ivadomed/model-seg-dcm/pull/10/commits/bdb2ac155cb53b64423d7b7cb8e416492629707e).

<img width="649" alt="image" src="https://github.com/ivadomed/model-seg-dcm/assets/39456460/11ba6eb3-7053-4511-bf0b-0f21dd481a65">

naga-karthik replied:
Thanks for the figures! maybe you should keep both `dropout_holes=False` and `dropout_holes=True` but with `fill_value=0`.

valosekj replied:
Good point! I'm now showing both `dropout_holes=False` and `dropout_holes=True` for both `fill_value=None` and `fill_value=0`:

<img width="632" alt="image" src="https://github.com/ivadomed/model-seg-dcm/assets/39456460/778081ae-d20d-43ed-b643-e2ce3ed638ff">

naga-karthik replied:
so we can go ahead with the `fill_value=0` set of transforms? what do you think?

valosekj replied:
Yes! I pushed `fill_value=0` in https://github.com/ivadomed/model-seg-dcm/pull/10/commits/bdb2ac155cb53b64423d7b7cb8e416492629707e. The model is currently training!

naga-karthik commented on vit_unetr_ssl/transforms.py:
```diff
@@ -0,0 +1,224 @@
+                        dropout_holes=True,
+                        # max_spatial_size=(roi_size[0]//4, roi_size[1]//4, roi_size[2]//4)
+                    ),
+                    # 'dropout_holes=False': the areas inside the holes will be filled with random noise
```
Likewise, this should be clarified as well! `dropout_holes=False` is just the opposite -- the `holes` will be kept and the rest of the image is dropped out 

Remember, we saw a full images with only the SC in the middel and the rest of the image was noise? So, I guess that the result of 2 successive transforms: (1) `dropout_holes=False` which drops out the entire image (except the holes) and fills the values with 0 (because the default value for `fill_value` is `0`) and (2) `RandCoarseShuffle` shuffles the pixels throughout the image.

valosekj replied:
Again, good catch! You are right. See [my comment above](https://github.com/ivadomed/model-seg-dcm/pull/10#discussion_r1527289602). I'll clarify the comments in future commits.

valosekj replied:
Fixed in https://github.com/ivadomed/model-seg-dcm/pull/10/commits/7f2734084cdb1e0cdaa4e4890992351e0873653d and https://github.com/ivadomed/model-seg-dcm/pull/10/commits/bdb2ac155cb53b64423d7b7cb8e416492629707e

naga-karthik commented on vit_unetr_ssl/transforms.py:
```diff
@@ -0,0 +1,224 @@
+                image_key="image",
+                image_threshold=0,
+            ),
+            # data-augmentation
+            # NOTE: the following transforms are based on contrast-agnostic data augmentation:
+            # https://github.com/ivadomed/model-seg-dcm/blob/nk/dcm-zurich-pretraining/monai/transforms.py
+            RandAffined(keys=["image", "label"], mode=(2, 1), prob=0.9,
+                        rotate_range=(-20. / 360 * 2. * np.pi, 20. / 360 * 2. * np.pi),
+                        # monai expects in radians
+                        scale_range=(-0.2, 0.2),
+                        translate_range=(-0.1, 0.1)),
+            Rand3DElasticd(keys=["image", "label"], prob=0.5,
+                           sigma_range=(3.5, 5.5),
+                           magnitude_range=(25., 35.)),
+            RandSimulateLowResolutiond(keys=["image"], zoom_range=(0.5, 1.0), prob=0.25),
+            RandAdjustContrastd(keys=["image"], gamma=(0.5, 3.), prob=0.5),  # this is monai's RandomGamma
+            RandBiasFieldd(keys=["image"], coeff_range=(0.0, 0.5), degree=3, prob=0.3),
+            RandGaussianNoised(keys=["image"], mean=0.0, std=0.1, prob=0.1),
+            RandGaussianSmoothd(keys=["image"], sigma_x=(0., 2.), sigma_y=(0., 2.), sigma_z=(0., 2.0),
+                                prob=0.3),
+            RandScaleIntensityd(keys=["image"], factors=(-0.25, 1), prob=0.15),
+            # this is nnUNet's BrightnessMultiplicativeTransform
+            RandFlipd(keys=["image", "label"], prob=0.3, ),
```
oh man, taking these transforms exactly from `contrast-agnostic` is not a good idea! They heavily deform the image (useful for SC seg but not for lesion seg). We need much simpler transforms for lesion seg. Add my suggestions below

valosekj replied:
Agree. Indeed, I just copy-pasted the transforms. But note that these transforms are for the fine-tuned model, which can theoretically be any type of segmentation (SC, lesions, ...). But agree, I will accept your suggestions! Thanks!

valosekj replied:
Committed in https://github.com/ivadomed/model-seg-dcm/pull/10/commits/a4182b5f91aa8693e617ae933fde3856fc8a2ad1

naga-karthik commented on vit_unetr_ssl/transforms.py:
```diff
@@ -0,0 +1,224 @@
+            # data-augmentation
+            # NOTE: the following transforms are based on contrast-agnostic data augmentation:
+            # https://github.com/ivadomed/model-seg-dcm/blob/nk/dcm-zurich-pretraining/monai/transforms.py
+            RandAffined(keys=["image", "label"], mode=(2, 1), prob=0.9,
+                        rotate_range=(-20. / 360 * 2. * np.pi, 20. / 360 * 2. * np.pi),
+                        # monai expects in radians
+                        scale_range=(-0.2, 0.2),
+                        translate_range=(-0.1, 0.1)),
+            Rand3DElasticd(keys=["image", "label"], prob=0.5,
+                           sigma_range=(3.5, 5.5),
+                           magnitude_range=(25., 35.)),
+            RandSimulateLowResolutiond(keys=["image"], zoom_range=(0.5, 1.0), prob=0.25),
+            RandAdjustContrastd(keys=["image"], gamma=(0.5, 3.), prob=0.5),  # this is monai's RandomGamma
+            RandBiasFieldd(keys=["image"], coeff_range=(0.0, 0.5), degree=3, prob=0.3),
+            RandGaussianNoised(keys=["image"], mean=0.0, std=0.1, prob=0.1),
+            RandGaussianSmoothd(keys=["image"], sigma_x=(0., 2.), sigma_y=(0., 2.), sigma_z=(0., 2.0),
+                                prob=0.3),
+            RandScaleIntensityd(keys=["image"], factors=(-0.25, 1), prob=0.15),
+            # this is nnUNet's BrightnessMultiplicativeTransform
+            RandFlipd(keys=["image", "label"], prob=0.3, ),
```
```suggestion
            RandAffined(keys=["image", "label"], mode=(2, 1), prob=0.1,
                        rotate_range=(-20. / 360 * 2. * np.pi, 20. / 360 * 2. * np.pi),
                        # monai expects in radians
                        scale_range=(-0.2, 0.2),
                        translate_range=(-0.1, 0.1)),
            RandSimulateLowResolutiond(keys=["image"], zoom_range=(0.5, 1.0), prob=0.2),
            RandAdjustContrastd(keys=["image"], gamma=(0.5, 3.), prob=0.2),  # this is monai's RandomGamma
            RandBiasFieldd(keys=["image"], coeff_range=(0.0, 0.5), degree=3, prob=0.1),
            RandGaussianSmoothd(keys=["image"], sigma_x=(0., 2.), sigma_y=(0., 2.), sigma_z=(0., 2.0),
                                prob=0.2),
            RandFlipd(keys=["image", "label"], prob=0.2, ),
```

valosekj committed:
Update transforms for training of the fine-tuned model

From contrast-agnostic SC seg transforms (which heavily deform the image) to simpler transforms more suitable for lesion seg.
Context: https://github.com/ivadomed/model-seg-dcm/pull/10#discussion_r1527186705

Co-authored-by: Naga Karthik Enamundram <53445351+naga-karthik@users.noreply.github.com>

valosekj committed:
update comment, remove unused imports

valosekj committed:
Add notebook with RandCoarseDropoutd transform debug

valosekj committed:
Fix 'dropout_holes=True' and 'dropout_holes=False' comments

From MONAI doc (https://docs.monai.io/en/latest/transforms.html#randcoarsedropoutd):
- if True, dropout the regions of holes and fill value specified by 'fill_value'
- if False, keep the holes and dropout the outside and fill value specified by 'fill_value'

valosekj committed:
remove unused 'max_spatial_size' arg

valosekj committed:
use 'fill_value=0' for 'RandCoarseDropoutd'

valosekj committed:
Plot also RandCoarseDropoutd dropout_holes=False fill_value=0

valosekj committed:
Add note that the batch size is actually doubled (8*2=16), because we are using two augmented samples per 3D Volume

valosekj committed:
Add '--cuda' input arg

valosekj committed:
Remove 'AsDiscrete'

valosekj committed:
Remove 'AsDiscrete'

valosekj committed:
Add 'CUDA_NUM=args.cuda'

valosekj committed:
Add TODO to increase batch_size to 16

valosekj committed:
Use 'roi_size' for 'define_finetune_train_transforms'

valosekj committed:
Use 'label_sc' to crop samples around the SC

- we have to do  it because the pretrained model was trained on samples of (64, 64, 64)

valosekj committed:
batch_size = 8

valosekj committed:
NUM_WORKERS = batch_size

valosekj committed:
Add 'import torch.multiprocessing'

valosekj committed:
Fix shape logging

valosekj committed:
Change 'img_size' to 'ROI_SIZE'

valosekj committed:
'batch["label"]' --> 'batch["label_lesion"]'

valosekj committed:
Plot and save input and output validation images to see how the model is learning

valosekj committed:
Fix ROI_SIZE for sliding_window_inference

valosekj committed:
Crop samples of 64x64x64 also for Validation of the fine-tuned model

valosekj committed:
update docstring

valosekj committed:
log validation samples shapes

valosekj committed:
Save validation figure only if it contains a lesion

valosekj committed:
Plot GT together with image

valosekj committed:
print unique values in the slice to see if it is binary

valosekj committed:
update output fig fname

valosekj committed:
Add 'AsDiscreted' for Training and  Validation of the fine-tuned model

valosekj committed:
threshold val_labels_list and val_outputs_list by 0.5 threshold before computing validation dice and plotting

naga-karthik committed:
add normalized relu normalization

naga-karthik committed:
fix binarization bug

valosekj committed:
remove 'logger.info(np.unique(output.detach().cpu().numpy()))'

valosekj committed:
overlay prediction over input image

valosekj committed:
Fix variable when getting probabilities from logits

valosekj committed:
Add debug lines

valosekj committed:
PEP8

valosekj committed:
Set validation batch_size to 1

valosekj committed:
improve comments

valosekj committed:
fix figure title

valosekj committed:
comment 'AsDiscreted' transforms

valosekj committed:
Make '--pretrained-model' non required to allow training from the scratch

valosekj committed:
Add script to create spine-generic MSD dataset

valosekj committed:
run notebook again

valosekj committed:
Make 'create_msd_data.py' compatible with other BIDS datasets

valosekj committed:
Add note that no testing set is created

valosekj committed:
Add README with instructions on how to download T2w images from multiple datasets

valosekj committed:
fix typo

valosekj committed:
remove unused imports

valosekj committed:
fix sc suffix for sci-paris

valosekj committed:
Use os.path.abspath for 'args.path_data' and 'args.path_out'

valosekj committed:
Update logging message

valosekj committed:
Fix '.replace' to prevent '//' in the output string

valosekj committed:
Add 'create_msd_data.py' commands to create MSD-style JSON datalists
